Python/compile.c

```diff
@@ -2065,6 +2079,72 @@ wrap_in_stopiteration_handler(struct compiler *c)
     return SUCCESS;
 }
 
+static int
+compiler_type_params(struct compiler *c, asdl_typeparam_seq *typeparams)
+{
+    if (!typeparams) {
+        return SUCCESS;
+    }
+    Py_ssize_t n = asdl_seq_LEN(typeparams);
+
+    for (Py_ssize_t i = 0; i < n; i++) {
+        typeparam_ty typeparam = asdl_seq_GET(typeparams, i);
+        location loc = LOC(typeparam);
+        switch(typeparam->kind) {
+        case TypeVar_kind:
+            ADDOP_LOAD_CONST(c, loc, typeparam->v.TypeVar.name);
+            if (typeparam->v.TypeVar.bound) {
+                expr_ty bound = typeparam->v.TypeVar.bound;
+                if (compiler_enter_scope(c, typeparam->v.TypeVar.name, COMPILER_SCOPE_TYPEPARAMS,
+                                        (void *)typeparam, bound->lineno) == -1) {
+                    return ERROR;
+                }
+                VISIT(c, expr, bound);
+                ADDOP_IN_SCOPE(c, loc, RETURN_VALUE);
+                PyCodeObject *co = optimize_and_assemble(c, 1);
+                compiler_exit_scope(c);
+                if (co == NULL) {
+                    return ERROR;
+                }
+                if (compiler_make_closure(c, loc, co, 0) < 0) {
+                    Py_DECREF(co);
+                    return ERROR;
+                }
+                Py_DECREF(co);
+                if (c->u->u_ste->ste_type_params_in_class) {
+                    ADDOP(c, loc, LOAD_LOCALS);
+                    ADDOP_I(c, loc, CALL_INTRINSIC_2, INTRINSIC_SET_CLASS_DICT);
+                }
+
+                int intrinsic = bound->kind == Tuple_kind
+                    ? INTRINSIC_TYPEVAR_WITH_CONSTRAINTS
+                    : INTRINSIC_TYPEVAR_WITH_BOUND;
```

Comment:
What do you think about removing this hardcoded "detect a tuple" from the compiler and doing it at runtime instead? One less intrinsic, and IMO a clearer boundary that the parser and compiler don't care what expression comes after the colon, only the runtime does.

Reply:
What I don't like about that is that it's not statically detectable whether we end up with a bound or a constraint (because even if bound is a single name, it might evaluate to a tuple at runtime). But I don't feel too strongly.

Reply:
Ah, good point. So the real question here is what should be the meaning of this code:

```
X = (int, str)
def f[T: X]():
    ...
```

Does `T` have an invalid bound, or is it constrained to `int` or `str`?

I think you're right that this must be an invalid bound, since static checkers can't interpret it as a constraint.